packages/app-admin-ui/client/providers/List.js

```diff
@@ -38,3 +38,19 @@ export const ListProvider = ({ list, children }) => {
     </ListContext.Provider>
   );
 };
+
+const ListDataContext = createContext();
+export const useListData = () => useContext(ListDataContext);
+
+export const ListDataProvider = ({ query, items, selectedItems, onSelectChange, children }) => (
```

Comment:
I'm reluctant to add another provider if it can at all be avoided. I see why it might be needed but wonder what other patterns you considered. One thing that is not clear to me is why `onSelectChange` is managed by the `ListDataProvider` but `handleDelete` isn't.

If we are going to have one, I'd like to think about what it should be responsible for. For example, can it handle fetching and re-fecting data and just provide events back? I'm not sure how it should work.

Reply:
on digging further I see that `query` and `items` can be used from `useList()`.

`onSelectChange` is the behaviour which  is passed down from `useListSelect()`, unfortunately this is not using s shared state, every time you have `useListSelect(items)` even though the items are same but the internal tracking of selection is different. I have tried using `useListSelect()` in components but they do not get selected items.

We can merge this with `useList()` and that should help consolidate this. let me try do that.

Reply:
https://github.com/keystonejs/keystone/pull/2482/commits/54b208e4f604f600f8b2cd63f16ba74bc1cbb0cd removes the extra ListDataProvider but the commit does not work due to `useListSelect` not being shared for list.

I had to merge `useListSelect` with `useList` so that `useList` now also returns `selectedItems` and `onSelectChange` (should be renamed to `setSelectedItems` really).